apps/examples/src/examples/speech-bubble/SpeechBubble/helpers.tsx

```diff
@@ -1,211 +1,158 @@
-						? new Vec2d(intersection.x - offset.horizontal, intersection.y)
-						: new Vec2d(intersection.x + offset.horizontal, intersection.y),
-					new Vec2d(handle.x, handle.y),
-					line === 0
-						? new Vec2d(intersection.x + offset.horizontal, intersection.y)
-						: new Vec2d(intersection.x - offset.horizontal, intersection.y),
-			  ]
-			: [
-					line === 1
-						? new Vec2d(intersection.x, intersection.y - offset.vertical)
-						: new Vec2d(intersection.x, intersection.y + offset.vertical),
-					new Vec2d(handle.x, handle.y),
-					line === 1
-						? new Vec2d(intersection.x, intersection.y + offset.vertical)
-						: new Vec2d(intersection.x, intersection.y - offset.vertical),
-			  ]
-	}
+	const { adjustedIntersection, intersectionSegmentIndex } = getHandleIntersectionPoint(shape)
 
-	let modifiedSegments = [...initialSegments]
+	let vertices: Vec2d[]
 
 	// Inject the tail segments into the geometry of the shape
-	switch (line) {
+	switch (intersectionSegmentIndex) {
 		case 0:
-			modifiedSegments.splice(1, 0, ...createTailSegments('horizontal'))
+			// top
+			vertices = [
+				tl,
+				new Vec2d(adjustedIntersection.x - offsetH, adjustedIntersection.y),
+				new Vec2d(handleInShapeSpace.x, handleInShapeSpace.y),
+				new Vec2d(adjustedIntersection.x + offsetH, adjustedIntersection.y),
+				tr,
+				br,
+				bl,
+			]
 			break
 		case 1:
-			modifiedSegments.splice(2, 0, ...createTailSegments('vertical'))
+			// right
+			vertices = [
+				tl,
+				tr,
+				new Vec2d(adjustedIntersection.x, adjustedIntersection.y - offsetV),
+				new Vec2d(handleInShapeSpace.x, handleInShapeSpace.y),
+				new Vec2d(adjustedIntersection.x, adjustedIntersection.y + offsetV),
+				br,
+				bl,
+			]
 			break
 		case 2:
-			modifiedSegments.splice(3, 0, ...createTailSegments('horizontal'))
+			// bottom
+			vertices = [
+				tl,
+				tr,
+				br,
+				new Vec2d(adjustedIntersection.x + offsetH, adjustedIntersection.y),
+				new Vec2d(handleInShapeSpace.x, handleInShapeSpace.y),
+				new Vec2d(adjustedIntersection.x - offsetH, adjustedIntersection.y),
+				bl,
+			]
 			break
 		case 3:
-			modifiedSegments = [...modifiedSegments, ...createTailSegments('vertical')]
+			// left
+			vertices = [
+				tl,
+				tr,
+				br,
+				bl,
+				new Vec2d(adjustedIntersection.x, adjustedIntersection.y + offsetV),
+				new Vec2d(handleInShapeSpace.x, handleInShapeSpace.y),
+				new Vec2d(adjustedIntersection.x, adjustedIntersection.y - offsetV),
+			]
 			break
 		default:
 			// eslint-disable-next-line no-console
-			console.log('default')
+			throw Error("no intersection found, this shouldn't happen")
 	}
 
-	return modifiedSegments
+	return vertices
 }
 
-export function getHandleIntersectionPoint({
-	w,
-	h,
-	handle,
-}: {
-	w: number
-	h: number
-	handle: TLHandle
-}) {
-	const offset = { horizontal: w / 10, vertical: h / 10 }
-	const handleVec = new Vec2d(handle.x, handle.y)
+export function getHandleIntersectionPoint(shape: SpeechBubbleShape) {
```

Comment:
I had a hard time reading through this function. My hunch was that by "flattening out" the helper functions, I could tell the story a bit easier. You'll hear other opinions on this kind of thing, but IMO a flatter function is more useful than splitting subroutines out into their own smaller named functions just for the sake of organization, like you had with `getAdjustedIntersectionPoint` and `checkIntersection`. (Sidebar, this is the opposite of how I feel about React components—split those fuckers out as much as you can!—but that's because unlike stuff like this, bits and pieces of UI tend to develop complexity over time and it's nice to have that complexity contained).

Reply:
Thanks for the tip! I definitely heard the opposite opinion a lot when starting out. Clarity should always be the goal though, and this way is clearer for sure.